substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JDKConfigFiles.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2023, 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.hosted.jdk;
+
+import com.oracle.svm.core.BuildArtifacts;
+import com.oracle.svm.core.feature.AutomaticallyRegisteredFeature;
+import com.oracle.svm.core.feature.InternalFeature;
+import com.oracle.svm.core.util.VMError;
+import org.graalvm.nativeimage.ImageSingletons;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+
+public final class JDKConfigFiles {
+
+    public static JDKConfigFiles singleton() {
+        return ImageSingletons.lookup(JDKConfigFiles.class);
+    }
+
+    final Set<String> configFiles = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    JDKConfigFiles() {
+    }
+
+    public void register(String configFile) {
+        if (configFile.startsWith("/")) {
+            configFiles.add(configFile.substring(1));
+        } else {
+            configFiles.add(configFile);
+        }
+    }
+}
+
+@AutomaticallyRegisteredFeature
+final class JDKConfigFilesFeature implements InternalFeature {
+    @Override
+    public void afterRegistration(AfterRegistrationAccess access) {
+        ImageSingletons.add(JDKConfigFiles.class, new JDKConfigFiles());
+    }
+
+    @Override
+    public void afterImageWrite(AfterImageWriteAccess access) {
+        Path jdkHomeDir = Path.of(System.getProperty("java.home")).normalize();
+        for (String configFile : new TreeSet<>(JDKConfigFiles.singleton().configFiles)) {
+            Path jdkConfigPath = jdkHomeDir.resolve(configFile).normalize();
+            VMError.guarantee(jdkConfigPath.startsWith(jdkHomeDir));
```

Comment:
Doesn't `resolve().normalize()` take care of this? If you want to keep the guarantee, please add a descriptive message.

Reply:
I still need to validate this, but the idea was to protect against passing in `../../../someSystemFile`, which after `.normalize()` would no longer contain the `..` directories, but would point to a fille outside of the `java.home` dir. That is unless I'm missing something.
Since it's not actually exposes to end user (will it ever be?) this validation is maybe a bit overly cautious.

Reply:
I agree that we probably don't need to be too cautious here as it's internal API. However, the validation again happens late in the process, I'd prefer if we validate such things as early as possible (in the `register()` hook in this case).